src/Widget/Widget.php

```diff
@@ -91,98 +92,44 @@ public static function end()
 
                 return $widget;
             }
-
-            throw new InvalidCallException('Expecting end() of ' . get_class($widget) . ', found ' . static::class);
+            throw new \BadFunctionCallException('Expecting end() of ' . get_class($widget) . ', found ' . static::class);
         }
-
-        throw new InvalidCallException('Unexpected ' . static::class . '::end() call. A matching begin() is not found.');
+        throw new \BadFunctionCallException(
+            'Unexpected ' . static::class . '::end() call. A matching begin() is not found.'
+        );
     }
 
     /**
-     * Creates a widget instance and runs it.
-     * The widget rendering result is returned by this method.
-     *
-     * @param array $config name-value pairs that will be used to initialize the object properties
-     *
-     * @return string the rendering result of the widget.
-     * @throws \Exception
+     * Creates a widget instance.
      *
+     * @return Widget $widget.
      */
-    public static function widget($config = []): string
+    public static function widget(): Widget
     {
-        ob_start();
-        ob_implicit_flush(false);
-
-        try {
-            /* @var $widget Widget */
-            $config['__class'] = static::class;
-            $widget = Yii::createObject($config);
-            $out = '';
-            if ($widget->beforeRun()) {
-                $result = $widget->run();
-                $out = $widget->afterRun($result);
-            }
-        } catch (\Exception $e) {
-            // close the output buffer opened above if it has not been closed already
-            if (ob_get_level() > 0) {
-                ob_end_clean();
-            }
+        $widget = new static(self::$eventDispatcher, self::$webView);
 
-            throw $e;
-        }
-
-        return ob_get_clean() . $out;
-    }
+        static::$widget = $widget;
 
-    private $id;
-
-    /**
-     * Returns the ID of the widget.
-     *
-     * @param bool $autoGenerate whether to generate an ID if it is not set previously
-     *
-     * @return string ID of the widget.
-     */
-    public function getId($autoGenerate = true): string
-    {
-        if ($autoGenerate && $this->id === null) {
-            $this->id = static::$autoIdPrefix . static::$counter++;
-        }
-
-        return $this->id;
+        return $widget;
     }
 
     /**
-     * Sets the ID of the widget.
+     * Returns the view object that can be used to render views or view files.
      *
-     * @param string $value id of the widget.
+     * The {@see render()} and {@see renderFile()} methods will use this view object to implement the actual view
+     * rendering. If not set, it will default to the "view" application component.
      */
-    public function setId(string $value): void
+    public function getView(): WebView
```

Comment:
Do widgets require `WebView` instead of `View`?

Reply:
If correct they require WebView since it is initialized in the application with ViewFactory.

Reply:
But WebView extends View so the question is would it work with View?

Reply:
Yes work fine with View also, but I think it should be with WebView since it is the one configured in ViewFactory